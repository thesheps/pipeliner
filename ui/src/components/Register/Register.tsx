import React, { useState } from "react";
import Button from "@material-ui/core/Button";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogContentText from "@material-ui/core/DialogContentText";
import DialogTitle from "@material-ui/core/DialogTitle";
import TextField from "@material-ui/core/TextField";

export interface RegisterProps {
  registerUser: (
    username: string,
    emailAddress: string,
    password: string
  ) => void;
}

export const Register = ({ registerUser }: RegisterProps) => {
  const [username, setUsername] = useState("");
  const [emailAddress, setEmailAddress] = useState("");
  const [emailError, setEmailError] = useState(false);
  const [password, setPassword] = useState("");
  const [open, setOpen] = useState(false);

  const validateEmailAddress = (email: string) => {
    const re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    const isValid = re.test(email.toLowerCase());

    setEmailAddress(email);

    if (!isValid) {
      setEmailError(true);
    } else {
      setEmailError(false);
    }
  };

  return (
    <>
      <Button
        data-testid="register-dialog-button"
        color="inherit"
        onClick={() => setOpen(true)}
      >
        Register
      </Button>

      <Dialog open={open} onClose={() => setOpen(false)}>
        <DialogTitle id="form-dialog-title">Register</DialogTitle>

        <DialogContent>
          <DialogContentText>Register here to get going!</DialogContentText>

          <TextField
            value={username}
            onChange={(e) => setUsername(e.target.value)}
            inputProps={{ "data-testid": "username-input" }}
            autoFocus
            margin="dense"
            label="Username"
            type="text"
            fullWidth
          />

          <TextField
            value={emailAddress}
            onChange={(e) => validateEmailAddress(e.target.value)}
            inputProps={{ "data-testid": "emailAddress-input" }}
            margin="dense"
            label="Email"
            type="email"
            error={!!emailError}
            helperText="Please enter a valid email address"
            fullWidth
          />

          <TextField
            value={password}
            onChange={(e) => setPassword(e.target.value)}
            inputProps={{ "data-testid": "password-input" }}
            margin="dense"
            label="Password"
            type="password"
            fullWidth
          />
        </DialogContent>

        <DialogActions>
          <Button
            data-testid="cancel-button"
            onClick={() => setOpen(false)}
            color="primary"
          >
            Cancel
          </Button>

          <Button
            data-testid="register-button"
            onClick={() => registerUser(username, emailAddress, password)}
            color="primary"
            disabled={!username || !emailAddress || !password || !!emailError}
          >
            Register
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};
