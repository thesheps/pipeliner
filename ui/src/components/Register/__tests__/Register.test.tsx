import React from "react";
import { fireEvent, render } from "@testing-library/react";
import {
  toBeDisabled,
  toBeVisible,
  toHaveStyle,
} from "@testing-library/jest-dom/matchers";

import { Register } from "../Register";

expect.extend({ toBeDisabled, toBeVisible, toHaveStyle });

describe("Register", () => {
  const username = "joe.bloggs";
  const emailAddress = "joe.bloggs@invalid.com";
  const password = "my super awesome password";

  it("renders correctly", () => {
    const { container } = render(<Register registerUser={jest.fn()} />);

    expect(container).toMatchSnapshot();
  });

  it("validates username has been entered", () => {
    const { getByText, getByTestId } = render(
      <Register registerUser={jest.fn()} />
    );

    fireEvent.click(getByText("Register"));

    fireEvent.change(getByTestId("emailAddress-input"), {
      target: { value: emailAddress },
    });

    fireEvent.change(getByTestId("password-input"), {
      target: { value: password },
    });

    expect(getByTestId("register-button")).toBeDisabled();
  });

  it("validates emailAddress has been entered", () => {
    const { getByText, getByTestId } = render(
      <Register registerUser={jest.fn()} />
    );

    fireEvent.click(getByText("Register"));

    fireEvent.change(getByTestId("username-input"), {
      target: { value: emailAddress },
    });

    fireEvent.change(getByTestId("password-input"), {
      target: { value: password },
    });

    expect(getByTestId("register-button")).toBeDisabled();
  });

  it("validates password has been entered", () => {
    const { getByText, getByTestId } = render(
      <Register registerUser={jest.fn()} />
    );

    fireEvent.click(getByText("Register"));

    fireEvent.change(getByTestId("emailAddress-input"), {
      target: { value: emailAddress },
    });

    fireEvent.change(getByTestId("username-input"), {
      target: { value: username },
    });

    expect(getByTestId("register-button")).toBeDisabled();
  });

  it("calls registerUser function upon clicking register", () => {
    const registerUser = jest.fn();
    const { getByText, getByTestId } = render(
      <Register registerUser={registerUser} />
    );

    fireEvent.click(getByText("Register"));

    fireEvent.change(getByTestId("username-input"), {
      target: { value: username },
    });

    fireEvent.change(getByTestId("emailAddress-input"), {
      target: { value: emailAddress },
    });

    fireEvent.change(getByTestId("password-input"), {
      target: { value: password },
    });

    fireEvent.click(getByTestId("register-button"));
    expect(registerUser).toHaveBeenCalledWith(username, emailAddress, password);
  });

  it("calls hide modal function upon clicking cancel", () => {
    const { getByText, getByTestId } = render(
      <Register registerUser={jest.fn()} />
    );

    fireEvent.click(getByText("Register"));
    fireEvent.click(getByTestId("cancel-button"));

    expect(getByTestId("cancel-button")).not.toBeVisible();
  });
});
