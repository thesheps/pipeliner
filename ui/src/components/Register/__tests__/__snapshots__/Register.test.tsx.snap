// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register renders correctly 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
    data-testid="register-dialog-button"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Register
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;
